clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java

```diff
@@ -269,11 +305,11 @@ public void testFilterToEmptyBatchRetention() {
 
                 ByteBuffer filtered = ByteBuffer.allocate(2048);
                 MemoryRecords.FilterResult filterResult = records.filterTo(new TopicPartition("foo", 0),
-                        new MemoryRecords.RecordFilter() {
+                        new MemoryRecords.RecordFilter(0, 0) {
                             @Override
-                            protected BatchRetention checkBatchRetention(RecordBatch batch) {
+                            protected BatchRetentionResult checkBatchRetention(RecordBatch batch) {
                                 // retain all batches
-                                return BatchRetention.RETAIN_EMPTY;
+                                return new BatchRetentionResult(BatchRetention.RETAIN_EMPTY, true);
```

Comment:
Since there is no marker, It seems that containsMarkerForEmptyTxn should be false.